File: rust/herb-linter/src/rules/html_img_require_alt.rs
use crate::utils::tag_utils::{element_tag_name_location, get_element_attribute, get_tag_local_name, has_attribute_value};

use herb::nodes::HTMLElementNode;
use herb::Visitor;

rule_visitor!(ImgRequireAltVisitor);
define_parser_rule!(
  HTMLImgRequireAltRule,
  "html-img-require-alt",
  Warning,
  ImgRequireAltVisitor,
  parser_options: { action_view_helpers: true },
  introduced_in: "0.4.0"
);

impl Visitor for ImgRequireAltVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if get_tag_local_name(node).as_deref() == Some("img") {
      match get_element_attribute(node, "alt") {
        None => self.add_offense(
          "Missing required `alt` attribute on `<img>` tag. Add `alt=\"\"` for decorative images or `alt=\"description\"` for informative images.",
          element_tag_name_location(node),
        ),

        Some(attribute) if !has_attribute_value(attribute) => self.add_offense(
          "The `alt` attribute has no value. Add `alt=\"\"` for decorative images or `alt=\"description\"` for informative images.",
          attribute.location.clone(),
        ),

        Some(_) => {}
      }
    }

    self.walk_html_element_node(node);
  }
}

File: javascript/packages/linter/src/rules/html-img-require-alt.ts
import { BaseRuleVisitor } from "./rule-utils.js"
import { hasAttribute, getAttribute, hasAttributeValue, getTagLocalName, isHTMLOpenTagNode, isERBOpenTagNode, filterHTMLAttributeNodes, findAttributeByName } from "@herb-tools/core"

import { ParserRule } from "../types.js"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, HTMLAttributeNode, ParseResult, ParserOptions } from "@herb-tools/core"

class ImgRequireAltVisitor extends BaseRuleVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    this.checkImgTag(node)
    super.visitHTMLElementNode(node)
  }

  private checkImgTag(node: HTMLElementNode): void {
    const tagName = getTagLocalName(node)

    if (tagName !== "img") {
      return
    }

    const altAttribute = this.getAltAttribute(node)

    if (!altAttribute) {
      this.addOffense(
        'Missing required `alt` attribute on `<img>` tag. Add `alt=""` for decorative images or `alt="description"` for informative images.',
        node.tag_name!.location
      )
      return
    }

    if (!hasAttributeValue(altAttribute)) {
      this.addOffense(
        'The `alt` attribute has no value. Add `alt=""` for decorative images or `alt="description"` for informative images.',
        altAttribute.location
      )
    }
  }

  private getAltAttribute(node: HTMLElementNode): HTMLAttributeNode | null {
    const openTag = node.open_tag

    if (isHTMLOpenTagNode(openTag)) {
      return getAttribute(openTag, "alt")
    }

    if (isERBOpenTagNode(openTag)) {
      return findAttributeByName(filterHTMLAttributeNodes(openTag.children), "alt")
    }

    return null
  }
}

export class HTMLImgRequireAltRule extends ParserRule {
  static ruleName = "html-img-require-alt"

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return { action_view_helpers: true }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new ImgRequireAltVisitor(this.ruleName, context)
    visitor.visit(result.value)
    return visitor.offenses
  }
}
